### ISSUE TITLE
Timestep distribution / shift: proposal

### ISSUE THREAD

**Issue by dxqb** [Lightricks/LTX-2#291]
I think there is an issue with how LTX determines the timestep shift (Flux-shift) both for inference and for training.
For inference, the shift calculation as shipped with the diffusers scheduler config lands at extreme values - that's why most pipeline set it to 7.7 instead.
For training, it's also not usable but corrected by your own training code by stretching the distribution back.

Here is a proposal to improve both:
**Abstract.** Flow-matching image and video models sample on a noise schedule whose shift must grow with resolution. The near-universal implementation, Flux-shift, fits a curve through two anchor points. It approximates SD3's resolution shift — a redundancy law: higher-resolution inputs carry redundant detail, so their schedule shifts toward higher noise, up to a physical ceiling where every added token is a perfect copy. For image models the fit is faithful; for video (LTX) it pushes past that ceiling and the shift diverges, which LTX papers over with ad-hoc training and inference corrections. We restore SD3's redundancy law with three changes: a tunable redundancy strength, independent spatial and temporal factors, and an anchor at a base resolution where the schedule is unshifted. The result reproduces Flux-shift for images inside their calibrated range, and removes the runaway for video that the LTX workarounds exist to patch.

---

## 1. Flux-shift: the law the schedulers ship

Flow-matching models train and sample on a noise schedule parameterized by a **shift**; higher shift pushes probability mass toward high noise. The shift that is right at one resolution is wrong at another, so schedulers make it resolution-dependent. The near-universal implementation — Flux's `calculate_shift`, inherited by diffusers and the LTX code — is what we call **Flux-shift**: it interpolates `mu` linearly in the **token count** between two anchors and exponentiates.

$$\mu = m \cdot L + b, \qquad m = \frac{S_\text{max} - S_\text{base}}{L_\text{max} - L_\text{base}}, \qquad \text{shift} = e^{\mu}$$

where *L* is the token count, and the base and max anchors are per-model token-count/shift pairs (for LTX: 1024 → 0.95 and 4096 → 2.05).

|        | Flux2                         | LTX                            |
|--------|-------------------------------|--------------------------------|
| `mu`   | `0.000169 * seq + 0.457`      | `0.000358 * seq + 0.583`       |
| base anchor | 256 tok → 1.649 (256²)   | 1024 tok → 2.586 (1024²)       |
| max anchor  | 4096 tok → 3.158 (1024²) | 4096 tok → 7.768 (2048²)       |
| floor `exp(b)` | 1.579                 | 1.792                          |

Two facts to carry forward. First, because the shift is `exp` of a straight line in the token count, it **accelerates**: the more tokens, the faster it climbs, and nothing bounds it. Second, it has no notion of an "unshifted" resolution: its lowest possible value is the floor `exp(b)` (1.58 / 1.79), never 1.0.

---

## 2. Where it comes from: SD3's resolution shift

The resolution dependence is not arbitrary. Higher-resolution images carry **redundant** information: neighbouring pixels are correlated, so adding tokens adds little *independent* signal. At a fixed noise level a larger image is effectively less noisy — you can average over the redundancy — so its schedule must be shifted toward higher noise to keep the per-independent-sample difficulty constant.

SD3 ("Scaling Rectified Flow Transformers for High-Resolution Image Synthesis", eq. 23) makes this quantitative. On the logit-SNR variable `t/(1-t)`, moving a schedule from `n` to `m` pixels is a multiplicative shift:

$$\frac{t_m}{1 - t_m} = \sqrt{\frac{m}{n}} \cdot \frac{t_n}{1 - t_n}$$

Two things matter:

- **It is relative.** The rule maps a schedule you already have at one resolution onto another; it carries no absolute level. You supply the schedule at the reference resolution.
- **`sqrt` is the redundancy ceiling.** It assumes every added token is a *perfect* copy: doubling tokens doubles the independent-sample-equivalents, so effective SNR doubles and the shift grows as its square root. You cannot beat a perfect copy.

**SD3 has the power law but no free exponent** — the `sqrt` is fixed — and it did not use that ceiling. SD3's high-resolution shift was chosen empirically from a human-preference study as `alpha = 3.0` at 1024². Measured against an unshifted 256² pretraining base (a 16× token ratio):

$$r = \frac{2\ln(\alpha)}{\ln(\text{token ratio})} = \frac{2\ln 3.0}{\ln 16} = 0.79$$

The physical `sqrt` (`r = 1`) would have given `16^0.5 = 4.0` at 1024²; SD3 chose `3.0`, i.e. it **backed off the ceiling to an effective redundancy exponent of 0.79.** That backoff is exactly what we will parameterize, and 0.79 is where our default comes from — read straight out of the paper.

So there are two laws in play: SD3's **power law** (a fixed-exponent rule — the `sqrt`, softened to `r ≈ 0.79` in practice), and **Flux-shift**, the `exp(linear-in-tokens)` approximation the schedulers actually run. The next section compares them.

---

## 3. The comparison, in plots

Two plots compare Flux-shift (red) against the redundancy law from an unshifted 256px base (green `r=0.8`; blue `r=1.0`, the physical ceiling). The **shaded interpolation range** in each is where Flux-shift interpolates between its two anchors rather than extrapolating.

**First, the shift each law produces at a given token count:**

<img width="1690" height="728" alt="Image" src="https://github.com/user-attachments/assets/a0b0087f-34fd-46ea-b32f-d8e4f9c76063" />

- **Flux2:** inside its interpolation range Flux-shift tracks the `r=0.8` power law closely — a faithful approximation of SD3's softened shift for an image model.
- **LTX:** Flux-shift pulls far above even the `r=1.0` physical law and keeps climbing — at 0.5 MP it reaches **~400 by 10 s**, while the redundancy law stays around 9–13. Real video sizes sit deep in its runaway extrapolation.

**Second, the redundancy each Flux-shift value implies** — the same shift restated as the exponent `r` it corresponds to, so it can be read against the ceiling. It diverges at the base, dips through the physical band, then climbs with token count:

<img width="1690" height="650" alt="Image" src="https://github.com/user-attachments/assets/81721621-c08e-4948-8a2c-66f604427535" />

- **Flux2:** stays in the physical band across its range, reaching **0.83 at its max anchor** (= SD3's `alpha=3.0`); crosses `r=1` only in extrapolation. Sane.
- **LTX:** reaches **0.99 at its max anchor** (right at the ceiling, §5.5) and goes **super-physical** just past it, hitting **~2.2** at 0.5 MP · 10 s — more than perfect redundancy, which is impossible.

**Takeaway:** Flux-shift faithfully approximates SD3's softened shift for images inside its calibrated range, but for video its implied redundancy passes the physical ceiling and the shift runs away. The next section shows what LTX had to do to live with that.

---

## 4. How LTX copes with the runaway

LTX ships Flux-shift and then works around its video failure in two different, ad-hoc ways — one in training, one in inference. Neither changes the law; both patch its symptoms.

### 4.1 In training: Flux-shift, then re-stretched

LTX's training timestep sampler (`ltx-trainer`, `ShiftedLogitNormalTimestepSampler`) starts from Flux-shift. `_get_shift_for_sequence_length` is exactly the linear `mu` law — same 1024/4096 → 0.95/2.05 anchors — and it is **unclamped**, so for video token counts `mu` extrapolates and grows without bound. Timesteps are then drawn as a shifted logit-normal, `t = sigmoid(N(mu, std))`. Because `mu` is large at video sizes, that distribution **collapses onto `t ≈ 1`** — the entire training budget lands on high noise, and nothing is left to learn the mid/low-noise regime.

Their fix is two arbitrary corrections layered on top:

1. **Percentile stretch.** They renormalize the samples between the 0.5th and 99.9th percentiles of the (collapsed) logit-normal back onto [0, 1], forcing coverage.
2. **10% uniform mix.** With probability 0.1 a sample is replaced by a uniform draw — the docstring states this "prevents collapse at high token counts."

<img width="1170" height="676" alt="Image" src="https://github.com/user-attachments/assets/680e91d5-09e3-4831-b744-b98f51f8f49e" />

The plot is the training timestep distribution at a common video size (0.5 MP · 5 s = 7808 tokens). The plain Flux-shift logit-normal (red, shift ≈ 29) is a spike on `t = 1` — a collapsed schedule. LTX's corrected sampler (purple) claws coverage back with the stretch and the visible flat 10% uniform pedestal. The redundancy law (green, shift ≈ 6.8) does not collapse: its shift stays in a sane range, so the plain logit-normal already spreads across the useful timesteps, and **the percentile stretch has nothing to undo**. The 10% uniform mix is a separate matter — a blanket coverage choice, orthogonal to the shift law; the redundancy law neither needs it nor replaces it, and it could be applied on top just the same. Of the two corrections, only the stretch exists to patch Flux-shift's runaway — and that one the redundancy law makes unnecessary.

### 4.2 In inference: pinned to a fixed shift

The inference pipelines sidestep the law entirely. Nearly every LTX-2 pipeline calls the scheduler *without a latent*, so `number_of_tokens` falls back to its default `MAX_SHIFT_ANCHOR = 4096` and the pipeline applies a fixed `exp(2.05) = 7.768` to everything, independent of shape.

<img width="1105" height="650" alt="Image" src="https://github.com/user-attachments/assets/0e676441-3961-4acf-954e-e35d092a55dc" />

At 0.5 MP, sweeping clip length: the pipeline (red) holds flat at 7.768 for every duration, because it always assumes 4096 tokens (≈ 2.5 s). Its interpolation range covers only ~0.4–2.5 s, so the common 5 s and 10 s clips are already extrapolation. A length-aware law has to *scale* with frame count; Flux-shift conflates all tokens into one number and cannot, so the pipeline just holds the shift constant.

The plot also exposes a limit of **our own** law. Its temporal factor grows with every added frame, so above ~7 s it climbs past the pipeline's 7.768 and keeps rising. That assumes the added frames are *redundant* — more of the same scene — which holds for the short single-shot clips LTX trains on, but not for long video full of cuts and new content, where the extra frames carry **independent** signal and the true temporal redundancy saturates. So for long, edited clips our constant `r_t` **over-shifts**; a fuller treatment would let `r_t` fall off with duration once the scene stops repeating.

---

## 5. Our contribution: back to SD3's law, lightly modified

The core of our contribution is not a new law — it is to **go back to SD3's original power law**, the one Flux-shift approximated and, for video, broke. Restoring it is most of the benefit. On top of that we make three slight modifications: expose its exponent as a tunable knob (SD3 fixed it at the `sqrt`), split it into independent spatial and temporal factors, and anchor it at a base resolution where the shift is 1.0.

### 5.1 The law

$$\text{shift} = \left(\frac{HW}{HW_\text{ref}}\right)^{r_s/2} \cdot \left(\frac{F}{F_\text{ref}}\right)^{r_t/2}$$

`HW`, `F` are the per-frame token count and **latent** frame count of the target (§5.4); `HW_ref`, `F_ref` the same at the base. The shift is **1.0 at the base shape** by construction — the schedule is unshifted at the pretraining resolution and only rises away from it. There is no separate "base shift" level; the base resolution *is* the anchor. (This is where Flux-shift differs even for images: its base anchor is elevated to 1.649 / 2.586, so it never expresses an unshifted base.)

### 5.2 What the exponent means

`r` is the redundancy exponent: SNR doublings gained per token doubling. `(token ratio) ** r` is the effective redundancy — the independent-sample-equivalents you could average.

- `r = 1`: SD3's derivation, the physical ceiling (perfect copies).
- `r = 0`: no resolution dependence, shift 1.0 everywhere.
- `0 < r < 1`: realistic; tokens partly, not perfectly, redundant. SD3's own empirical choice is `r ≈ 0.79`.

### 5.3 Why spatial and temporal separate

Redundancy is multiplicative across independent axes. A longer clip of one scene is nearly a redundant *temporal* copy — a different redundancy from spatial neighbours — so each axis gets its own exponent, and they separate into two factors. For an image model `F = F_ref = 1` and the temporal factor is a no-op, recovering SD3's pure spatial rule.

### 5.4 Why the reference is in pixels

`HW_ref` is a pixel resolution, converted to tokens with each model's own compression ratios. A 1024² frame is 4096 tokens on Flux2 (16 px/token) but 1024 tokens on LTX (32 px/token). The law only uses *ratios*, so the differing absolute counts cancel as long as reference and target use the same model's conversion — one pixel-valued reference is model-agnostic.

This ratio-invariance holds for the spatial axis because px→token is a *proportional* rescale (through the origin). It does **not** hold for the temporal axis: the VAE's frame packing is affine (e.g. LTX `latent = (frames − 1) / 8 + 1`), so pixel- and latent-frame ratios differ. `F` must therefore be counted in **latent frames** — matching `HW`, which is already post-patchify tokens — not pixel frames.

### 5.5 Chosen defaults

- **`reference_resolution = 256`** (px): where the law sets shift = 1.0. For Flux2 this is principled — SD3 pretrains at 256² and its `alpha=3.0` study fixes an unshifted 256² base (a least-squares fit to Flux2's Flux-shift also lands at ≈240px). LTX has no such base — it trains multi-resolution with no single pretraining resolution (LTX-Video, arXiv:2501.00103) — so for LTX 256px is a convention, kept only to have one model-agnostic reference (§5.4).
- **`r_s = r_t = 0.8`**: SD3's empirical redundancy exponent (0.79, rounded), below the `r=1` ceiling because tokens are only partly redundant. We keep the two axes equal for now — the split (§5.3) lets them differ, but we have no firm basis to push the temporal exponent above or below the spatial one: latent frames are ~⅓ s apart (8× compression), so they are neither independent nor near-copies.

---

## 6. Summary

| | Flux-shift (shipped) | this work |
|---|---|---|
| functional form | `exp(linear in token count)` | power law in token ratio |
| base | none; floor 1.58 (Flux2) / 1.79 (LTX), never 1.0 | shift 1.0 at pretraining resolution |
| parameters | two anchor shifts, per model | `r_s`, `r_t`, base resolution (model-agnostic) |
| spatial/temporal | conflated (token count only) | independent exponents |
| implied redundancy, image (Flux2) | ~0.8 in range — **works** | 0.8 (= SD3's `alpha=3.0`) |
| implied redundancy, video (LTX) | super-physical past the anchor — **breaks** | tunable, `r ≤ 1` |
| coping in training | Flux-shift + percentile stretch + 10% uniform | none needed; shift stays in range |
| coping in inference | pinned flat to 7.768 | scales with resolution and length |
| past the max anchor | runs away exponentially | scales as a fixed power |

The redundancy law recovers SD3's derivation at `r = 1`, adopts SD3's own empirical exponent (0.79) as its default, reproduces Flux-shift for image models inside their calibrated range, stays near the shift LTX's working inference pipeline applies, and removes the reason for LTX's training and inference workarounds — the runaway they exist to patch.

**Comment by michaellightricks** [comment]
Thanks for the detailed write-up.

The current inference shift override and the training-time stretch of the schedule are intentional. A redesign of the Flux-shift / redundancy-law schedule is not on the near-term roadmap, but it may be considered in the future.

Closing this here.
